ledger/accountdb.go

```diff
@@ -486,17 +508,106 @@ func makeCompactResourceDeltas(accountDeltas []ledgercore.NewAccountDeltas, base
 				if appParams.Params != nil {
 					newEntry.newResource.SetAppParams(*appParams.Params, false)
 				} else {
-					newEntry.newResource.ClearAppParams()
+					// ClearAppLocalState sets resourceFlagsNotHolding flag.
+					// Otherwise newEntry.newResource.ResourceFlags == resourceFlagsHolding (0)
+					// meaning empty asset hodling (or app local state)
+					newEntry.newResource.ClearAppLocalState()
 				}
 				if baseResourceData, has := baseResources.read(appParams.Addr, basics.CreatableIndex(appParams.Aidx)); has {
 					newEntry.oldResource = baseResourceData
 					outResourcesDeltas.insert(appParams.Addr, basics.CreatableIndex(appParams.Aidx), newEntry) // insert instead of upsert economizes one map lookup
 				} else {
 					outResourcesDeltas.insertMissing(appParams.Addr, basics.CreatableIndex(appParams.Aidx), newEntry)
+					missingAddresses[appParams.Addr] = struct{}{}
 				}
 			}
 		}
 	}
+
+	outResourcesDeltas.knownAddresses = make(map[basics.Address]int64, len(missingAddresses))
```

Comment:
I'm not sure if having the `knownAddresses` / `missingAddresses` is the best route here; what would happen if we were to try and store the `addrid` on each of the entries in `newEntry.oldResource`, and leave it with zero if we don't know that from the cache ? i.e. in `resourcesLoadOld` we would compare it against zero and attempt to load it as needed. Instead of having the `missingAddresses`, I would add the address to the `resourcesDeltas`

Reply:
good idea, done